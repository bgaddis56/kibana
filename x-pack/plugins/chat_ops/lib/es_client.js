/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
import { Client } from 'elasticsearch';
//import { es } from '../config.json';
const client = new Client({ host: 'localhost:9200',
  httpAuth: 'elastic:changeme'
});

export default client;
