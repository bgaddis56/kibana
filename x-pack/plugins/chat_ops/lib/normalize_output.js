/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
export default function normalizeOutput(node) {
  if (typeof node === 'object') {
    if (!node.type) throw new Error('Objects must have a type propery');
    if (!node.value) throw new Error('Objects must have a value property');
    return node;
  }

  return { type: typeof node, value: node };
}
