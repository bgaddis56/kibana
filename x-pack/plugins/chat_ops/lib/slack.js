/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
import Slack from 'slack';

export default (server) => {
  const config = server.config();
  const token = config.get('xpack.chatops.chattoken');

  const slack = new Slack({ token });
  return slack;
};
