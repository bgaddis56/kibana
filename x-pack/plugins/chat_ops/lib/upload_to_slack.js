/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
import { WebClient } from '@slack/client';

export async function uploadToSlack(server, options) {
  const config = server.config();
  const chattoken = config.get('xpack.chatops.chattoken');

  const payload = { ...options, token: chattoken };
  const inputBot = new WebClient(chattoken);

  return await inputBot.files.upload(payload);
}
