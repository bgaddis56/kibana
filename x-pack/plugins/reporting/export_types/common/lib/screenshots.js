/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

//import * as Rx from 'rxjs';
//import { first, tap, mergeMap } from 'rxjs/operators';
import fs from 'fs';
import getPort from 'get-port';
import { promisify } from 'bluebird';
import { LevelLogger } from '../../../server/lib/level_logger';

const fsp = {
  readFile: promisify(fs.readFile, fs)
};

export function screenshotsObservableFactory(server) {
  const config = server.config();
  const logger = LevelLogger.createForServer(server, ['reporting', 'screenshots']);

  const browserDriverFactory = server.plugins.reporting.browserDriverFactory;
  const captureConfig = config.get('xpack.reporting.capture');

  const asyncDurationLogger = async (description, promise) => {
    const start = new Date();
    const result = await promise;
    logger.debug(`${description} took ${new Date() - start}ms`);
    return result;
  };

  const openUrl = async (browser, url, headers) => {
    const waitForSelector = '.application';

    await browser.open(url, {
      headers,
      waitForSelector,
    });
  };

  const injectCustomCss = async (browser, layout) => {
    const filePath = layout.getCssOverridesPath();
    const buffer = await fsp.readFile(filePath);
    await browser.evaluate({
      fn: function (css) {
        const node = document.createElement('style');
        node.type = "text/css";
        node.innerHTML = css; // eslint-disable-line no-unsanitized/property
        document.getElementsByTagName('head')[0].appendChild(node);
      },
      args: [buffer.toString()],
    });
  };

  const waitForElementOrItemsCountAttribute = async (browser, layout) => {
    // the dashboard is using the `itemsCountAttribute` attribute to let us
    // know how many items to expect since gridster incrementally adds panels
    // we have to use this hint to wait for all of them
    await browser.waitForSelector(`${layout.selectors.renderComplete},[${layout.selectors.itemsCountAttribute}]`);
  };

  const waitForNotFoundError = async (browser) => {
    await browser.waitForSelector(`.toast.alert.alert-danger`);
    throw new Error('Reporting subject could not be loaded to take a screenshot.');
  };

  const getNumberOfItems = async (browser, layout) => {
    // returns the value of the `itemsCountAttribute` if it's there, otherwise
    // we just count the number of `itemSelector`
    const itemsCount = await browser.evaluate({
      fn: function (selector, countAttribute) {
        const elementWithCount = document.querySelector(`[${countAttribute}]`);
        if (elementWithCount) {
          return parseInt(elementWithCount.getAttribute(countAttribute));
        }

        return document.querySelectorAll(selector).length;
      },
      args: [layout.selectors.renderComplete, layout.selectors.itemsCountAttribute],
    });
    return itemsCount;
  };

  const waitForElementsToBeInDOM = async (browser, itemsCount, layout) => {
    await browser.waitFor({
      fn: function (selector) {
        return document.querySelectorAll(selector).length;
      },
      args: [layout.selectors.renderComplete],
      toEqual: itemsCount
    });
  };

  const setViewport = async (browser, itemsCount, layout) => {
    const viewport = layout.getViewport(itemsCount);
    await browser.setViewport(viewport);
  };

  const positionElements = async (browser, layout) => {
    if (layout.positionElements) {
      await layout.positionElements(browser);
    }
  };

  const waitForRenderComplete = async (browser, layout) => {
    await browser.evaluate({
      fn: function (selector, visLoadDelay) {
        // wait for visualizations to finish loading
        const visualizations = document.querySelectorAll(selector);
        const visCount = visualizations.length;
        const renderedTasks = [];

        function waitForRender(visualization) {
          return new Promise(function (resolve) {
            visualization.addEventListener('renderComplete', () => resolve());
          });
        }

        function waitForRenderDelay() {
          return new Promise(function (resolve) {
            setTimeout(resolve, visLoadDelay);
          });
        }

        for (let i = 0; i < visCount; i++) {
          const visualization = visualizations[i];
          const isRendered = visualization.getAttribute('data-render-complete');

          if (isRendered === 'disabled') {
            renderedTasks.push(waitForRenderDelay());
          } else if (isRendered === 'false') {
            renderedTasks.push(waitForRender(visualization));
          }
        }

        // The renderComplete fires before the visualizations are in the DOM, so
        // we wait for the event loop to flush before telling reporting to continue. This
        // seems to correct a timing issue that was causing reporting to occasionally
        // capture the first visualization before it was actually in the DOM.
        const hackyWaitForVisualizations = () => new Promise(r => setTimeout(r, 100));

        return Promise.all(renderedTasks).then(hackyWaitForVisualizations);
      },
      args: [layout.selectors.renderComplete, captureConfig.loadDelay],
      awaitPromise: true,
    });
  };

  const getTimeRange = async (browser, layout) => {
    const timeRange = await browser.evaluate({
      fn: function (fromAttribute, toAttribute) {
        const fromElement = document.querySelector(`[${fromAttribute}]`);
        const toElement = document.querySelector(`[${toAttribute}]`);

        if (!fromElement || !toElement) {
          return null;
        }

        const from = fromElement.getAttribute(fromAttribute);
        const to = toElement.getAttribute(toAttribute);
        if (!to || !from) {
          return null;
        }

        return { from, to };
      },
      args: [layout.selectors.timefilterFromAttribute, layout.selectors.timefilterToAttribute],
      returnByValue: true,
    });
    return timeRange;
  };

  const getElementPositionAndAttributes = async (browser, layout) => {
    const elementsPositionAndAttributes = await browser.evaluate({
      fn: function (selector, attributes) {
        const elements = document.querySelectorAll(selector);

        // NodeList isn't an array, just an iterator, unable to use .map/.forEach
        const results = [];
        for (const element of elements) {
          const boundingClientRect = element.getBoundingClientRect();
          results.push({
            position: {
              boundingClientRect: {
                // modern browsers support x/y, but older ones don't
                top: boundingClientRect.y || boundingClientRect.top,
                left: boundingClientRect.x || boundingClientRect.left,
                width: boundingClientRect.width,
                height: boundingClientRect.height,
              },
              scroll: {
                x: window.scrollX,
                y: window.scrollY
              }
            },
            attributes: Object.keys(attributes).reduce((result, key) => {
              const attribute = attributes[key];
              result[key] = element.getAttribute(attribute);
              return result;
            }, {})
          });
        }
        return results;

      },
      args: [layout.selectors.screenshot, { title: 'data-title', description: 'data-description' }],
      returnByValue: true,
    });

    return elementsPositionAndAttributes;
  };

  const getScreenshots = async ({ browser, elementsPositionAndAttributes }) => {
    const screenshots = [];
    for (const item of elementsPositionAndAttributes) {
      const base64EncodedData = await asyncDurationLogger('screenshot', browser.screenshot(item.position));

      screenshots.push({
        base64EncodedData,
        title: item.attributes.title,
        description: item.attributes.description
      });
    }
    return screenshots;
  };

  return async function screenshotsObservable(url, headers, layout, browserTimezone, cancellationToken) {

    const bridgePort = await getPort();

    logger.debug(`Creating browser driver factory`);

    const { browser, chromium } = await browserDriverFactory.create({
      bridgePort,
      viewport: layout.getBrowserViewport(),
      zoom: layout.getBrowserZoom(),
      logger,
      browserTimezone
    });

    if (cancellationToken) {
      cancellationToken.on(() => {
        logger.debug(`Closing Chromium due to cancellation`);
        chromium.close();
      });
    }

    logger.debug('Driver factory created');

    logger.debug(`opening ${url}`);
    await openUrl(browser, url, headers);

    logger.debug('injecting custom css');
    await injectCustomCss(browser, layout);

    logger.debug('waiting for elements or items count attribute; or not found to interrupt');
    Promise.race([waitForElementOrItemsCountAttribute(browser, layout), waitForNotFoundError(browser)])
      .then(function () {
      });

    logger.debug('determining how many items we have');
    const itemsCount = await getNumberOfItems(browser, layout);

    logger.debug('setting viewport');
    await setViewport(browser, itemsCount, layout);

    logger.debug(`waiting for ${itemsCount} to be in the DOM`);
    await waitForElementsToBeInDOM(browser, itemsCount, layout);

    logger.debug('positioning elements');
    await positionElements(browser, layout);

    logger.debug('waiting for rendering to complete');
    await  waitForRenderComplete(browser, layout);

    logger.debug('rendering is complete');
    const timeRange = await getTimeRange(browser, layout);

    logger.debug(timeRange ? `timeRange from ${timeRange.from} to ${timeRange.to}` : 'no timeRange');
    const elementsPositionAndAttributes = await getElementPositionAndAttributes(browser, layout);

    logger.debug(`taking screenshots`);
    const screenshots = await getScreenshots({ browser, elementsPositionAndAttributes });

    await chromium.close();

    return screenshots;
  };
}
