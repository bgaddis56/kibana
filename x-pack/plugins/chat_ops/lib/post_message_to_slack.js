/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
import slack from './slack';

export default (server, to, text, params = {}) => {

  const config = server.config();
  const chattoken = config.get('xpack.chatops.chattoken');
  const name = config.get('xpack.chatops.chatname');

  return slack(server).chat.postMessage({
    chattoken,
    text,
    username: name,
    as_user: true,
    channel: to,
    ...params,
  });
};
