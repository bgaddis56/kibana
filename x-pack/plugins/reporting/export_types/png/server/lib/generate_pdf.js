/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import * as Rx from 'rxjs';
import { toArray, mergeMap } from 'rxjs/operators';
import moment from 'moment';
import { groupBy } from 'lodash';
import { oncePerServer } from '../../../../server/lib/once_per_server';
import { screenshotsObservableFactory } from './screenshots';
import { getLayoutFactory } from './layouts';

const getTimeRange = (urlScreenshots) => {
  const grouped = groupBy(urlScreenshots.map(u => u.timeRange));
  const values = Object.values(grouped);
  if (values.length === 1) {
    return values[0][0];
  }

  return null;
};

const formatDate = (date, timezone) => {
  return moment.tz(date, timezone).format('llll');
};

function generatePdfObservableFn(server) {
  const screenshotsObservable = screenshotsObservableFactory(server);
  const captureConcurrency = 1;
  const getLayout = getLayoutFactory(server);

  const urlScreenshotsObservable = (urls, headers, layout) => {
    return Rx.from(urls).pipe(
      mergeMap(url => screenshotsObservable(url, headers, layout),
        (outer, inner) => inner,
        captureConcurrency
      )
    );
  };

  const createPdfWithScreenshots = async ({ title, browserTimezone, urlScreenshots }) => {
    if (title) {
      const timeRange = getTimeRange(urlScreenshots);
      title += (timeRange) ? ` â€” ${formatDate(timeRange.from, browserTimezone)} to ${formatDate(timeRange.to, browserTimezone)}` : '';
    }
    return urlScreenshots[0].screenshots[0].base64EncodedData;
  };


  return function generatePdfObservable(title, urls, browserTimezone, headers, layoutParams, logo) {
    const layoutParams1 = { dimensions: { height: 2808, width: 1227 },
      id: "preserve_layout"
    };
    const layout = getLayout(layoutParams1);
    const screenshots$ = urlScreenshotsObservable(urls, headers, layout);

    return screenshots$.pipe(
      toArray(),
      mergeMap(urlScreenshots => createPdfWithScreenshots({ title, browserTimezone, urlScreenshots, logo }))
    );
  };
}

export const generatePdfObservableFactory = oncePerServer(generatePdfObservableFn);
