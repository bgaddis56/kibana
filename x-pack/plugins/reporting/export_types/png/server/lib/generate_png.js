/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { oncePerServer } from '../../../../server/lib/once_per_server';
import { PreserveLayout } from '../../../common/layouts/preserve_layout';
import { screenshotsObservableFactory } from '../../../common/lib/screenshots';

function generatePngPromiseFn(server) {
  const screenshotsObservable = screenshotsObservableFactory(server);

  const urlScreenshot = async (url, headers, layout, browserTimezone, cancellationToken) => {

    const screenShots = await screenshotsObservable(url, headers, layout, browserTimezone, cancellationToken);

    if (screenShots.length !== 1) {
      throw new Error(`Expected there to be 1 screenshot, but there are ${screenShots.length}`);
    }

    return screenShots[0];

  };

  return async function generatePngPromise(url, browserTimezone, headers, layoutParams, cancellationToken) {

    if (!layoutParams || !layoutParams.dimensions) {
      throw new Error(`LayoutParams.Dimensions is undefined.`);
    }

    const layout =  new PreserveLayout(layoutParams.dimensions);

    const screenshot = await urlScreenshot(url, headers, layout, browserTimezone, cancellationToken);

    return {
      content_type: 'image/png',
      content: screenshot.base64EncodedData
    };

  };

}

export const generatePngPromiseFactory = oncePerServer(generatePngPromiseFn);
