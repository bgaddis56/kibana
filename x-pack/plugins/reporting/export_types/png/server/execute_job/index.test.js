/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import * as Rx from 'rxjs';
import { memoize } from 'lodash';
import { cryptoFactory } from '../../../../server/lib/crypto';
import { executeJobFactory } from './index';
import { generatePngObservableFactory } from '../lib/generate_png';

jest.mock('../lib/generate_png', () => ({ generatePngObservableFactory: jest.fn() }));

const cancellationToken = {
  on: jest.fn()
};

let mockServer;
let config;
beforeEach(() => {
  config = {
    'xpack.security.cookieName': 'sid',
    'xpack.reporting.encryptionKey': 'testencryptionkey',
    'xpack.reporting.kibanaServer.protocol': 'http',
    'xpack.reporting.kibanaServer.hostname': 'localhost',
    'xpack.reporting.kibanaServer.port': 5601,
    'server.basePath': '/sbp'
  };

  mockServer = {
    expose: () => { },
    config: memoize(() => ({ get: jest.fn() })),
    plugins: {
      elasticsearch: {
        getCluster: memoize(() => {
          return {
            callWithRequest: jest.fn()
          };
        })
      },
      security: null,
    },
    savedObjects: {
      getScopedSavedObjectsClient: jest.fn(),
    },
    uiSettingsServiceFactory: jest.fn().mockReturnValue({ get: jest.fn() }),
  };

  mockServer.config().get.mockImplementation((key) => {
    return config[key];
  });

  generatePngObservableFactory.mockReturnValue(jest.fn());
});

afterEach(() => generatePngObservableFactory.mockReset());

const encrypt = async (headers) => {
  const crypto = cryptoFactory(mockServer);
  return await crypto.encrypt(headers);
};

describe(`sessionCookie`, () => {
  test(`Fails if no relativeURL is passed in`, async () => {
    const executeJob = executeJobFactory(mockServer);
    const encryptedHeaders = await encrypt({});

    const generatePngObservable = generatePngObservableFactory();
    generatePngObservable.mockReturnValue(Rx.of(Buffer.from('')));

    await expect(executeJob({ headers: encryptedHeaders }, cancellationToken))
      .rejects
      .toThrowErrorMatchingSnapshot();
  });

  test(`if serializedSession doesn't exist it doesn't pass sessionCookie to generatePngObservable`, async () => {
    mockServer.plugins.security = {};
    const headers = {};
    const encryptedHeaders = await encrypt(headers);

    const generatePngObservable = generatePngObservableFactory();
    generatePngObservable.mockReturnValue(Rx.of(Buffer.from('')));

    const executeJob = executeJobFactory(mockServer);
    await executeJob({ relativeUrl: '/app/kibana#/something', headers: encryptedHeaders, session: null }, cancellationToken);

    expect(generatePngObservable).toBeCalledWith('http://localhost:5601/sbp/app/kibana#/something', undefined, null, undefined);
  });

  test(`if uses xpack.reporting.kibanaServer.hostname for domain of sessionCookie passed to generatePngObservable`, async () => {
    const sessionCookieOptions = {
      httpOnly: true,
      name: 'foo',
      path: '/bar',
      secure: false,
    };
    mockServer.plugins.security = {
      getSessionCookieOptions() {
        return sessionCookieOptions;
      },
    };
    const headers = {};
    const encryptedHeaders = await encrypt(headers);

    const session = 'thisoldesession';
    const encryptedSession = await encrypt(session);

    const generatePngObservable = generatePngObservableFactory();
    generatePngObservable.mockReturnValue(Rx.of(Buffer.from('')));

    const executeJob = executeJobFactory(mockServer);
    await executeJob({ relativeUrl: '/app/kibana#/something', headers: encryptedHeaders, session: encryptedSession }, cancellationToken);

    expect(generatePngObservable).toBeCalledWith('http://localhost:5601/sbp/app/kibana#/something', undefined, {
      domain: config['xpack.reporting.kibanaServer.hostname'],
      httpOnly: sessionCookieOptions.httpOnly,
      name: sessionCookieOptions.name,
      path: sessionCookieOptions.path,
      sameSite: 'Strict',
      secure: sessionCookieOptions.secure,
      value: session
    }, undefined);
  });

  test(`if uses server.host and reporting config isn't set for domain of sessionCookie passed to generatePngObservable`, async () => {
    config['xpack.reporting.kibanaServer.hostname'] = undefined;
    config['server.host'] = 'something.com';
    const sessionCookieOptions = {
      httpOnly: true,
      name: 'foo',
      path: '/bar',
      secure: false,
    };
    mockServer.plugins.security = {
      getSessionCookieOptions() {
        return sessionCookieOptions;
      },
    };
    const headers = {};
    const encryptedHeaders = await encrypt(headers);

    const session = 'thisoldesession';
    const encryptedSession = await encrypt(session);

    const generatePngObservable = generatePngObservableFactory();
    generatePngObservable.mockReturnValue(Rx.of(Buffer.from('')));

    const executeJob = executeJobFactory(mockServer);
    await executeJob({ relativeUrl: '/app/kibana#/something', headers: encryptedHeaders, session: encryptedSession }, cancellationToken);

    expect(generatePngObservable).toBeCalledWith('http://something.com:5601/sbp/app/kibana#/something', undefined, {
      domain: config['server.host'],
      httpOnly: sessionCookieOptions.httpOnly,
      name: sessionCookieOptions.name,
      path: sessionCookieOptions.path,
      sameSite: 'Strict',
      secure: sessionCookieOptions.secure,
      value: session
    }, undefined);
  });
});

test(`passes browserTimezone to generatePng`, async () => {
  const encryptedHeaders = await encrypt({});

  const generatePngObservable = generatePngObservableFactory();
  generatePngObservable.mockReturnValue(Rx.of(Buffer.from('')));

  const executeJob = executeJobFactory(mockServer);
  const browserTimezone = 'UTC';
  await executeJob({ relativeUrl: '/app/kibana#/something', browserTimezone, headers: encryptedHeaders }, cancellationToken);

  expect(generatePngObservable).toBeCalledWith('http://localhost:5601/sbp/app/kibana#/something', browserTimezone, null, undefined);
});

test(`adds forceNow to hash's query, if it exists`, async () => {
  const encryptedHeaders = await encrypt({});

  const generatePngObservable = generatePngObservableFactory();
  generatePngObservable.mockReturnValue(Rx.of(Buffer.from('')));

  const executeJob = executeJobFactory(mockServer);
  const forceNow = '2000-01-01T00:00:00.000Z';

  await executeJob({ relativeUrl: '/app/kibana#/something', forceNow, headers: encryptedHeaders }, cancellationToken);

  expect(generatePngObservable).toBeCalledWith('http://localhost:5601/sbp/app/kibana#/something?forceNow=2000-01-01T00%3A00%3A00.000Z', undefined, null, undefined);
});

test(`appends forceNow to hash's query, if it exists`, async () => {
  const encryptedHeaders = await encrypt({});

  const generatePngObservable = generatePngObservableFactory();
  generatePngObservable.mockReturnValue(Rx.of(Buffer.from('')));

  const executeJob = executeJobFactory(mockServer);
  const forceNow = '2000-01-01T00:00:00.000Z';

  await executeJob({
    relativeUrl: '/app/kibana#/something?_g=something',
    forceNow,
    headers: encryptedHeaders
  }, cancellationToken);

  expect(generatePngObservable).toBeCalledWith('http://localhost:5601/sbp/app/kibana#/something?_g=something&forceNow=2000-01-01T00%3A00%3A00.000Z', undefined, null, undefined);
});

test(`doesn't append forceNow query to url, if it doesn't exists`, async () => {
  const encryptedHeaders = await encrypt({});

  const generatePngObservable = generatePngObservableFactory();
  generatePngObservable.mockReturnValue(Rx.of(Buffer.from('')));

  const executeJob = executeJobFactory(mockServer);

  await executeJob({ relativeUrl: '/app/kibana#/something', headers: encryptedHeaders }, cancellationToken);

  expect(generatePngObservable).toBeCalledWith('http://localhost:5601/sbp/app/kibana#/something', undefined, null, undefined);
});

test(`returns content_type of image/png`, async () => {
  const executeJob = executeJobFactory(mockServer);
  const encryptedHeaders = await encrypt({});

  const generatePngObservable = generatePngObservableFactory();
  generatePngObservable.mockReturnValue(Rx.of(Buffer.from('')));

  const { content_type: contentType } = await executeJob({ relativeUrl: '/app/kibana#/something',
    timeRange: {}, headers: encryptedHeaders }, cancellationToken);
  expect(contentType).toBe('image/png');
});

test(`returns content of generatePng getBuffer base64 encoded`, async () => {
  const testContent = 'test content';

  const generatePngObservable = generatePngObservableFactory();
  generatePngObservable.mockReturnValue(Rx.of(Buffer.from(testContent)));

  const executeJob = executeJobFactory(mockServer);
  const encryptedHeaders = await encrypt({});
  const { content } = await executeJob({ relativeUrl: '/app/kibana#/something',
    timeRange: {}, headers: encryptedHeaders }, cancellationToken);

  expect(content).toEqual(Buffer.from(testContent).toString('base64'));
});
