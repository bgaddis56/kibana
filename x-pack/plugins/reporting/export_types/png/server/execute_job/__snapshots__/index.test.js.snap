// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sessionCookie Fails if no relativeURL is passed in 1`] = `"Unable to generate report. Url is not defined."`;
